dace/libraries/onnx/op_implementations/array_ops.py

```diff
@@ -0,0 +1,813 @@
+class EinsumTranspose(ONNXForward):
+
+    @staticmethod
+    def forward(node: onnx_op.ONNXOp, state: SDFGState, sdfg: SDFG) -> typing.Union[Node, SDFG]:
+        from dace.libraries.onnx.nodes.onnx_op_registry import ONNXEinsum  # avoid import loop
+        perm = node.perm
+        input_desc = in_desc_with_name(node, state, sdfg, "data")
+        output_desc = out_desc_with_name(node, state, sdfg, "transposed")
+
+        letters = [chr(ord('z') - i) for i in range(26)]
+        input_letters = "".join(letters[i] for i, _ in enumerate(input_desc.shape))
+        output_letters = "".join(letters[i] for i in perm)
+        equation_str = f"{input_letters}->{output_letters}"
+
+        nsdfg = dace.SDFG(node.label + "_expansion")
+        nstate = nsdfg.add_state()
+        einsum_node: onnx_op.ONNXOp = ONNXEinsum(node.label + "_einsum_expansion", equation=equation_str)
+
+        nstate.add_node(einsum_node)
+        einsum_node.add_in_connector("Inputs__0")
+        nsdfg.add_datadesc("data", copy.deepcopy(input_desc))
+        nsdfg.add_datadesc("transposed", copy.deepcopy(output_desc))
+        nsdfg.arrays["data"].transient = False
+        nsdfg.arrays["transposed"].transient = False
+
+        nstate.add_edge(nstate.add_read("data"), None, einsum_node, "Inputs__0", nsdfg.make_array_memlet("data"))
+        nstate.add_edge(einsum_node, "Output", nstate.add_write("transposed"), None,
+                        nsdfg.make_array_memlet("transposed"))
+
+        return nsdfg
+
+
+# ==============================================================================
+# Reshape and Flatten Operations
+# ==============================================================================
+
+
+@python_pure_op_implementation(shape=lambda reshaped: reshaped.shape,
+                               allowzero=lambda node: getattr(node, 'allowzero', 0))
+def Reshape(data, reshaped):
+    # If allowzero is 0 (default), we use numpy's reshape which doesn't allow zeros
+    # If allowzero is 1, we need to handle zeros in the shape tensor
+    if allowzero == 0:
+        reshaped[:] = np.reshape(data, shape)
+    else:
+        # For allowzero=1, we need to handle zeros in the shape tensor
+        # This means we need to preserve the original dimension size when a zero is encountered
+        new_shape = list(shape)
+        for i, dim in enumerate(new_shape):
+            if dim == 0:
+                new_shape[i] = data.shape[i]
+        reshaped[:] = np.reshape(data, new_shape)
+
+
+@python_pure_op_implementation(shape=lambda input, node: [prod(input.shape[:node.axis]), prod(input.shape[node.axis:])])
+def Flatten(input, output):
+    output[:] = input.reshape(shape)
+
+
+# ==============================================================================
+# Slicing Operations
+# ==============================================================================
+
+
+@op_implementation(op="Slice", name="pure")
+class PureSlice(ONNXForward):
```

Comment:
can we not implement this more simply with np.take_axis or something similar? (low prio)

Reply:
Please add to TODOs in project